alternator/executor.cc

```diff
@@ -957,21 +982,39 @@ future<executor::request_return_type> executor::list_tags_of_resource(client_sta
     return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
 }
 
-static void verify_billing_mode(const rjson::value& request) {
-        // Alternator does not yet support billing or throughput limitations, but
+struct billing_mode_type {
+    bool provisioned = false;
+    int rcu;
+    int wcu;
+};
+
+static billing_mode_type verify_billing_mode(const rjson::value& request) {
+    // Alternator does not yet support billing or throughput limitations, but
     // let's verify that BillingMode is at least legal.
     std::string billing_mode = get_string_attribute(request, "BillingMode", "PROVISIONED");
     if (billing_mode == "PAY_PER_REQUEST") {
         if (rjson::find(request, "ProvisionedThroughput")) {
             throw api_error::validation("When BillingMode=PAY_PER_REQUEST, ProvisionedThroughput cannot be specified.");
         }
     } else if (billing_mode == "PROVISIONED") {
-        if (!rjson::find(request, "ProvisionedThroughput")) {
+        const rjson::value *provisioned_throughput = rjson::find(request, "ProvisionedThroughput");
+        if (!provisioned_throughput) {
             throw api_error::validation("When BillingMode=PROVISIONED, ProvisionedThroughput must be specified.");
         }
+        const rjson::value& throughput = *provisioned_throughput;
+        auto rcu = get_int_attribute(throughput, "ReadCapacityUnits");
+        auto wcu = get_int_attribute(throughput, "WriteCapacityUnits");
+        if (!rcu.has_value()) {
+            throw api_error::validation("Value null at 'provisionedThroughput.readCapacityUnits' failed to satisfy constraint: Member must not be null");
```

Comment:
In my opinion, this is a really ugly error message and I would have changed it. We don't have to be compatible with the exact wording of the DynamoDB error message (just the type is critical!), and we usually aren't. It's just nice to have some common substring in the error message (e.g., "eadCapacityUnits", without the uncapitalized first letter!) to use in the raises() statement in tests.

Reply:
I wasn't sure how closely you wanted to be to DynamoDB. In a sense, it was easy to copy their message. It's also easy when testing, but I can create something else that would work just as well with a regex